Physick & Garratt Roughness Sublayer Model

akshaysridhar opened the pull request:
## To-do
- [x] Numerical integration (e.g. use QuadGK) in solution (currently DeRidder's analytical approximation is used) 
- [ ] Tests consistent with PG95 literature. 
- [x] Code structure following dispatch patterns used elsewhere within SurfaceFluxes

<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [x] I have read and checked the items on the review checklist.

charleskawczynski commented on Project.toml:
```diff
@@ -4,8 +4,10 @@ authors = ["Climate Modeling Alliance"]
 version = "0.5.0"
 
 [deps]
+CLIMAParameters = "6eacf6c3-8458-43b9-ae03-caf5306d3d53"
```
Just a reminder: no packages directly depend on CLIMAParameters anymore. We've only been adding dependence in the test suites

kmdeck replied:
@charleskawczynski maybe we could discuss this when you're back next week over slack! (ClimaLSM still does depend on CLIMAParameters)

charleskawczynski replied:
So, we've since changed directions, and high level packages are now depending on CLIMAParameters. They should still not be needed to use the package, but only needed for helper-like functionality.

LenkaNovak commented on src/test_pg_1995.jl:
```diff
@@ -0,0 +1,131 @@
+using Plots
```
This test is great! :) I have a few points:
- I've run it locally and the theta profiles for the RSL look somewhat off (negative and very different to the no-RSL case). The u profiles look fine - at least from a  qualitative comparison with Fig. 6.8 in Bonan's textbook.
- While it's nice to have the plot test somewhere, shall we add some automated value checking test for CI? 
- I presume the plan is to move this test from `/src`?

akshaysridhar replied:
Yes - plan is to move this into `test` and have more quantitative measures (with generated profiles for example)

LenkaNovak reviewed (commented):
Hi Akshay, this looks good. I left a few comments. Mostly minor, but I did notice the included test potentially produces some unexpected results. Lmk if you want to connect offline.

akshaysridhar commented on src/SurfaceFluxes.jl:
```diff
@@ -903,8 +905,71 @@ function recover_profile(
+# TODO: add tests
+# TODO: Verify that all current RSL models fall into this general code pattern, then 𝜙 can be abstracted
+"""
+function recover_profile_canopy(
```
Now that it's clear from the docs that all RSL models fall into this general framework we can use multiple dispatch on the existing profile recovery function.

akshaysridhar replied:
The revised files dispatch over  `CanopyType`(s), with `NoCanopy` `SparseCanopy` and `DenseCanopy` subtypes

akshaysridhar commented on src/SurfaceFluxes.jl:
```diff
@@ -903,8 +905,71 @@ function recover_profile(
+  - d = Displacement height (measure of the spatial lengthscale of the effect of the canopy roughness on near-wall turbulence)
+  - Z: Z coordinate(s) (within surface layer) for which variable values are required
+  - X_in,X_sfc: For variable X, values at interior and surface nodes
+  - transport: Transport type, (e.g. Momentum or Heat, used to determine physical scale coefficients)
```
Yes; the design documentation indicates that we assume tracers and heat have the same transport characteristics. I'll update the docstrings to reflect this.

charleskawczynski reviewed (commented):
Can we reduce the code duplication in this PR a bit? It'll make maintenance easier.

charleskawczynski reviewed (changes requested):
Please remove the .svg figures from this PR. Also, we'll want to at least squash the commits such that the .svg files do not appear in the commit history.

charleskawczynski commented:
They can be generated for the docs, but they should not be checked into the repo.

charleskawczynski commented on src/SurfaceFluxes.jl:
```diff
@@ -903,8 +930,95 @@ function recover_profile(
+) where {FT}
+    z_star = ca.z_star
+    d = ca.d
+    @assert isless.(Z - d, z_in(sc))
```
This assertion looks incorrect-- any broadcasted object is not going to return a Bool

charleskawczynski replied:
Maybe `all()` is missing?

charleskawczynski commented on src/SurfaceFluxes.jl:
```diff
@@ -903,8 +930,95 @@ function recover_profile(
+    d = ca.d
+    @assert isless.(Z - d, z_in(sc))
+    @show Z, d
+    @assert isless.(d, Z)
```
Same here

akshaysridhar commented:
@gdecker1 Can you please modify the plot outcomes such that they are generated within the docs ? (See, for example, the current implementation of the Universal Function plots that appear in the `docs/build` directory as a result of this approach using `plot_universal_functions.jl`)

akshaysridhar committed:
modified:   Project.toml
	modified:   docs/bibliography.bib
	modified:   docs/make.jl
	new file:   docs/src/SampleProfiles.md
	new file:   docs/src/plot_profiles.jl
	modified:   docs/src/plot_universal_functions.jl
	modified:   src/SurfaceFluxes.jl
	modified:   test/runtests.jl
	modified:   test/test_profiles.jl
	new file:   test/test_rsl.jl
	modified:   test/test_universal_functions.jl

akshaysridhar committed:
rel_tol test

akshaysridhar committed:
Merge remote-tracking branch 'origin/main' into as/PG95-rsl

	src/SurfaceFluxes.jl
	test/runtests.jl
	test/test_profiles.jl
	test/test_universal_functions.jl
